server/eclipse-project/src/main/java/us/freeandfair/corla/auth/AbstractAuthentication.java

```diff
@@ -0,0 +1,237 @@
+            result = false;
+          }
+        } else if (auth_stage.equals(AuthenticationInterface.SECOND_FACTOR)) {
+          // Or are we half-way through it?
+          if (secondFactorAuthenticate(the_request,
+                                       the_username,
+                                       the_second_factor)) {
+            // We have both traditionally and second-factor authenticated.
+            final Administrator admin = 
+                AdministratorQueries.byUsername(the_username);
+            admin.updateLastLoginTime();
+            Persistence.saveOrUpdate(admin);
+            the_request.session().attribute(AuthenticationInterface.AUTH_STAGE, 
+                                            AuthenticationInterface.ADMIN);
+            the_request.session().attribute(AuthenticationInterface.ADMIN, 
+                                            admin);
+            Main.LOGGER.info("Second factor authentication succeeded for administrator " + 
+                the_username);
+          } else {
+            // Send the authentication state machine back to its initial state.
+            the_request.session().attribute(AuthenticationInterface.AUTH_STAGE, 
+                                            AuthenticationInterface.PASSWORD);
+            Main.LOGGER.info("Second factor authentication failed for administrator" + 
+                the_username);
+            result = false;
+          }
+        }
+      }
+    } catch (final PersistenceException e) {
+      // there's nothing we can really do here other than saying that the
+      // authentication failed; it's also possible we failed to update the last
+      // login time, but that's not critical
+      deauthenticate(the_request, the_username);
+    }
+
+    if (!result) {
+      // a failed authentication attempt removes any existing session authentication 
+      deauthenticate(the_request, the_username);
+      Main.LOGGER.info("Authentication failed for user " + the_username);
+    }
+
+    return result;
+  }
+  
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public County authenticatedCounty(final Request the_request) {
+    County result = null;
+    if (isAuthenticatedAs(the_request, AdministratorType.COUNTY,
+                    the_request.queryParams(AuthenticationInterface.USERNAME))) {
+      final Administrator admin = 
+          (Administrator) the_request.session().attribute(ADMIN);
+      if (admin != null) {
+        result = CountyQueries.forAdministrator(admin);
+      }
+    }
+    return result;
+  }
+  
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void traditionalDeauthenticate(final Request the_request,
+                                        final String the_username) {
+    the_request.session().removeAttribute(ADMIN);
+    Main.LOGGER.info("session is now traditionally deauthenticated");
+  }
+  
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void twoFactorDeauthenticate(final Request the_request,
```

Comment:
I'm not certain what the need is for two different types of deauthentication. Are they meant to be overridden by child classes? It seems like it would be odd to let child classes mess with the session handling.

Reply:
I'm also not sure why the `deauthenticate` method needs a username parameter. I always thought of it as "deauthenticate the session", so it doesn't really matter who the session was authenticated as.

Reply:
The design is driven by the fact that we may someday support:
 - being authenticated to multiple roles simultaneously, and
 - some authentication systems force one to identify the credential being de/un-authenticated, though many do not

And yes, the two different kinds of authentication and de-authentication are meant to be overridden by child classes.  You'll see that in the Entrust stuff it matters.